shared/utility/motion/ForwardKinematics.h

```diff
@@ -324,61 +318,198 @@ namespace motion {
+                std::pair<Eigen::Vector3d, double>{Eigen::Vector3d::Zero(), 0.0},
+            };
+
+            // Add the torso into the list of particles, if requested
+            if (include_torso) {
+                particles[15] = std::pair<Eigen::Vector3d, double>{
+                    Eigen::Vector3d{model.massModel.torso.x(), model.massModel.torso.y(), model.massModel.torso.z()},
+                    model.massModel.torso.w()};
             }
 
-            if (includeTorso) {
-                arma::vec4 massVector;
-                for (size_t i = 0; i < 4; i++) {
-                    massVector[i] = model.massModel.masses[20][i];
-                }
-                // NUClear::log<NUClear::DEBUG>("calculateCentreOfMass - reading mass Torso", massVector);
-                double jointMass = massVector[3];
-                utility::math::matrix::Transform3D massScaler;
-                massScaler.submat(0, 0, 2, 2) *= jointMass;
-                totalMassVector += massScaler * massVector;  // = m * local centre of mass in global robot coords
+            // Calculate the CoM for the entire robot
+            std::pair<Eigen::Vector3d, double> robot_com = {Eigen::Vector3d::Zero(), 0.0};
+            for (const auto& particle : particles) {
+                robot_com.first = (robot_com.first * robot_com.second + particle.first * particle.second)
+                                  / (robot_com.second + particle.second);
+                robot_com.second = robot_com.second + particle.second;
             }
 
-            utility::math::matrix::Transform3D normaliser;
-            if (totalMassVector[3] > 0) {
-                normaliser.submat(0, 0, 2, 2) *= 1 / totalMassVector[3];
-                return normaliser * totalMassVector;
+            return Eigen::Vector4d{robot_com.first.x(), robot_com.first.y(), robot_com.first.z(), robot_com.second};
+        }  // namespace kinematics
+
+        /*! @brief Transforms inertial tensors for each robot particle into torso space and sums to find the total
+           inertial tensor
+            @return [[xx, xy, xz], relative to the torso basis
+                     [xy, yy, yz],
+                     [xz, yz, zz]]
+        */
+        inline Eigen::Matrix3d calculateInertialTensor(
+            const message::motion::KinematicsModel& model,
+            const std::array<Eigen::Matrix<double, 4, 4, Eigen::DontAlign>, 20>& forward_kinematics,
+            bool include_torso) {
+
+            // Convenience function to transform particle-space inertial tensors to torso-space inertial tensor
+            // Htp - transform from particle space to torso space
+            // particle - CoM coordinates in particle space
+            auto translateTensor =
+                [](const Eigen::Matrix4d& Htp, const Eigen::Matrix3d& tensor, const Eigen::Vector4d& com_mass) {
+                    Eigen::Vector4d com;
+                    com << com_mass.x(), com_mass.y(), com_mass.z(), 1.0;
```

Comment:
Just use the constructor `Eigen::Vector4d com(com_mass.x(), com_mass.y(), com_mass.z(), 1.0);`